### ISSUE TITLE
Inferencing after .h5 -> .pb conversion

### ISSUE THREAD

**Issue by fbarilla** [fizyr/keras-retinanet#1222]
I successfuly retrained a Retinanet network against my own dataset. Everything is working as expected when I run it against a video stream. Object are detected but performances are poor on a Jetson Xavier so I'm trying to move it under TensorRT and it's where I started to be confused :-) 
I converted the .h5 to .pb and used the TF/NVidia tooling to optimize the model. 
With Keras, getting the bounding box and displaying them is straightforward:

        boxes, scores, labels = model.predict_on_batch(np.expand_dims(image, axis=0))

With the Tensorflow model I use a construct like below:

input_names = ['input_1']
output_names = ['classification/concat','regression/concat']
tf_input = tf_sess.graph.get_tensor_by_name(input_names[0] + ':0')
tf_input.shape.as_list()

tf_classification = tf_sess.graph.get_tensor_by_name(output_names[1] + ':0')
tf_regression = tf_sess.graph.get_tensor_by_name(output_names[0] + ':0')

regression, classification = tf_sess.run([tf_classification, tf_regression], feed_dict={
    tf_input: image[None, ...]
})

'regression' and 'classification' come back populated with the right shape. The question now is: 
HOW DO I MAP 'regression' and 'classification' TO BOUNDING BOXES, SCORES, LABELS ?

**Comment by hgaiser** [comment]
The TensorRT model is the converted retinanet model? Then it should have bboxes, labels and scores already as an output. You seem to be taking the raw classification and regression outputs, which you can do if you want, but then you will have to interpret them yourself.

**Comment by fbarilla** [comment]
Hi Hans, thanks for the answer. The idea is to optimize the Retinanet model to run on edge devices (Jetson TX2, Xavier).  So after conversion to TF model I exercised different optimization schema (quantization, node pruning, FP16, INT8, ....). My confusion went from the fact that under TF (session.run) I was using the default graph outputs (regression, classification). I did not realized that the boxes, scores and labels are directly available through the 'filtered_detections/map/TensorArrayStack[_1/_2]/TensorArrayGatherV3' tensors. Getting values directly out of these tensors fixed the problem.
Maybe would be good to document it for user who want to optimized the Retinanet model under TF.   
   
Regards,Franck BarillaudExt Phone: (512) 286-5242 Tie Line: 363-5242e-mail: fbarilla@us.ibm.com
 
 
----- Original message -----From: Hans Gaiser <notifications@github.com>To: fizyr/keras-retinanet <keras-retinanet@noreply.github.com>Cc: fbarilla <fbarilla@us.ibm.com>, Author <author@noreply.github.com>Subject: [EXTERNAL] Re: [fizyr/keras-retinanet] Inferencing after .h5 -> .pb conversion (#1222)Date: Thu, Dec 19, 2019 2:39 AM 
The TensorRT model is the converted retinanet model? Then it should have bboxes, labels and scores already as an output. You seem to be taking the raw classification and regression outputs, which you can do if you want, but then you will have to interpret them yourself.

**Comment by guruzen** [comment]
Can u share the code to convert h5 to pb pls?

**Comment by hgaiser** [comment]
Thanks for the update @fbarilla. Since we're not officially supporting this, I don't think we should put it in the documentation as it would indicate we do support it.

**Comment by muralikrishnat29** [comment]
> Hi Hans, thanks for the answer. The idea is to optimize the Retinanet model to run on edge devices (Jetson TX2, Xavier).  So after conversion to TF model I exercised different optimization schema (quantization, node pruning, FP16, INT8, ....). My confusion went from the fact that under TF (session.run) I was using the default graph outputs (regression, classification). I did not realized that the boxes, scores and labels are directly available through the 'filtered_detections/map/TensorArrayStack[_1/_2]/TensorArrayGatherV3' tensors. Getting values directly out of these tensors fixed the problem. Maybe would be good to document it for user who want to optimized the Retinanet model under TF.        Regards,Franck BarillaudExt Phone: (512) 286-5242 Tie Line: 363-5242e-mail: fbarilla@us.ibm.com     ----- Original message -----From: Hans Gaiser <notifications@github.com>To: fizyr/keras-retinanet <keras-retinanet@noreply.github.com>Cc: fbarilla <fbarilla@us.ibm.com>, Author <author@noreply.github.com>Subject: [EXTERNAL] Re: [fizyr/keras-retinanet] Inferencing after .h5 -> .pb conversion (#1222)Date: Thu, Dec 19, 2019 2:39 AM  The TensorRT model is the converted retinanet model? Then it should have bboxes, labels and scores already as an output. You seem to be taking the raw classification and regression outputs, which you can do if you want, but then you will have to interpret them yourself. —You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

Hi @fbarilla, could you please let me know how did you use the outputs from filter_detections nodes to convert to bounding box and to denote in the image?

**Comment by fbarilla** [comment]
Take a look at the two files attached, (Keras, TF). Hope you'll find your 
answers:

 

Regards,
Franck Barillaud
STSM, CTO ISV Power Cloud Infrastructure
Master Inventor
Ext Phone: (512) 286-5242    Tie Line: 363-5242
e-mail: fbarilla@us.ibm.com

**Comment by MathieuHaller** [comment]
@fbarilla Hello, I'm also currently optimizing a keras-retinanet model, and I'm curious on how you converted your TF model to FP16 using tensorRT. 

Did you use TF-TRT ? If so, did you encounter any issues with custom layers from retinanet ? And finally did it make your model faster ?

Cheers !

**Comment by fbarilla** [comment]
here is the jupyter notebook that does the optimization. 



Sorry I did not run any 'real' performance tests, so I don't any any 
answer for your question.

Regards,
Franck Barillaud
STSM, CTO ISV Power Cloud Infrastructure
Master Inventor
Ext Phone: (512) 286-5242    Tie Line: 363-5242
e-mail: fbarilla@us.ibm.com

**Comment by rithvikrthampy** [comment]
Hi can you also include me in the loop i am unable to find the file the the above comments

**Comment by fbarilla** [comment]
Regards,
Franck Barillaud
STSM, CTO ISV Power Cloud Infrastructure
Master Inventor
Ext Phone: (512) 286-5242    Tie Line: 363-5242
e-mail: fbarilla@us.ibm.com

**Comment by rithvikrthampy** [comment]
@fbarilla i am still not able to get the code. I will send you a message in mail can you share it to me via mail?
